pkg/controller/machinepool/ibmcloudactuator.go

```diff
@@ -0,0 +1,138 @@
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/openshift/hive/pkg/ibmclient"
+)
+
+// IBMCloudActuator encapsulates the pieces necessary to be able to generate
+// a list of MachineSets to sync to the remote cluster
+type IBMCloudActuator struct {
+	logger    log.FieldLogger
+	ibmClient ibmclient.API
+}
+
+var _ Actuator = &IBMCloudActuator{}
+
+func addIBMCloudProviderToScheme(scheme *runtime.Scheme) error {
+	return machineapi.AddToScheme(scheme)
+}
+
+// NewIBMCloudActuator is the constructor for building an IBMCloudActuator
+func NewIBMCloudActuator(ibmCreds *corev1.Secret, scheme *runtime.Scheme, logger log.FieldLogger) (*IBMCloudActuator, error) {
+	ibmClient, err := ibmclient.NewClientFromSecret(ibmCreds)
+	if err != nil {
+		logger.WithError(err).Warn("failed to create IBM client with creds in clusterDeployment's secret")
+		return nil, err
+	}
+	actuator := &IBMCloudActuator{
+		logger:    logger,
+		ibmClient: ibmClient,
+	}
+	return actuator, nil
+}
+
+// GenerateMachineSets satisfies the Actuator interface and will take a clusterDeployment and return a list of MachineSets
+// to sync to the remote cluster.
+func (a *IBMCloudActuator) GenerateMachineSets(cd *hivev1.ClusterDeployment, pool *hivev1.MachinePool, logger log.FieldLogger) ([]*machineapi.MachineSet, bool, error) {
+	if cd.Spec.ClusterMetadata == nil {
+		return nil, false, errors.New("ClusterDeployment does not have cluster metadata")
+	}
+	if cd.Spec.Platform.IBMCloud == nil {
+		return nil, false, errors.New("ClusterDeployment is not for IBMCloud")
+	}
+	if pool.Spec.Platform.IBMCloud == nil {
+		return nil, false, errors.New("MachinePool is not for IBMCloud")
+	}
+
+	computePool := baseMachinePool(pool)
+	computePool.Platform.IBMCloud = &installertypesibmcloud.MachinePool{
+		InstanceType: pool.Spec.Platform.IBMCloud.InstanceType,
+		Zones:        pool.Spec.Platform.IBMCloud.Zones,
+	}
+
+	if pool.Spec.Platform.IBMCloud.DedicatedHosts != nil {
+		dedicatedHosts := []installertypesibmcloud.DedicatedHost{}
+		for _, host := range pool.Spec.Platform.IBMCloud.DedicatedHosts {
+			dedicatedHosts = append(dedicatedHosts, installertypesibmcloud.DedicatedHost{
+				Name:    host.Name,
+				Profile: host.Profile,
+			})
+		}
+		computePool.Platform.IBMCloud.DedicatedHosts = dedicatedHosts
+	}
+
+	if pool.Spec.Platform.IBMCloud.BootVolume != nil {
+		computePool.Platform.IBMCloud.BootVolume = &installertypesibmcloud.BootVolume{
+			EncryptionKey: pool.Spec.Platform.IBMCloud.BootVolume.EncryptionKey,
+		}
+	}
+
+	if len(computePool.Platform.IBMCloud.Zones) == 0 {
+		zones, err := a.ibmClient.GetVPCZonesForRegion(context.TODO(), cd.Spec.Platform.IBMCloud.Region)
+		if err != nil {
+			return nil, false, errors.Wrap(err, "compute pool not providing list of zones and failed to fetch list of zones")
+		}
+		if len(zones) == 0 {
+			return nil, false, fmt.Errorf("zero zones returned for region %s", cd.Spec.Platform.IBMCloud.Region)
```

Comment:
This makes me wonder if we ought to validate a _provided_ list of Zones against those the API thinks exist for the specified Region. (Here and also for [aws](https://github.com/openshift/hive/blob/98aed49eda6a4ab455b9a0aed8be7232a53d5705/pkg/controller/machinepool/awsactuator.go#L148), [azure](https://github.com/openshift/hive/blob/98aed49eda6a4ab455b9a0aed8be7232a53d5705/pkg/controller/machinepool/azureactuator.go#L76), [gcp](https://github.com/openshift/hive/blob/98aed49eda6a4ab455b9a0aed8be7232a53d5705/pkg/controller/machinepool/gcpactuator.go#L217).)